go/mysql/fastparse/fastparse.go

```diff
@@ -45,6 +45,22 @@ func ParseUint64(s string, base int) (uint64, error) {
 		i++
 	}
 
+	if i >= uint(len(s)) {
+		return 0, fmt.Errorf("cannot parse uint64 from %q", s)
+	}
+	// For some reason, MySQL parses things as uint64 even with
+	// a negative sign and then turns it into the 2s complement value.
+	minus := s[i] == '-'
+	if minus {
+		if !allowNeg {
+			return 0, fmt.Errorf("cannot parse uint64 from %q", s)
+		}
+		i++
+		if i >= uint(len(s)) {
+			return 0, fmt.Errorf("cannot parse uint64 from %q", s)
+		}
+	}
```

Comment:
Very nice catch. Also:

![image](https://github.com/vitessio/vitess/assets/42793/d9d11e35-d61f-4b8b-bcc7-5d7ee1d29442)

Also: I think the `bool` flag to the API makes it a bit sharp/footgun, and it makes the body confusing. Could you please split this this in two functions? I think there's a pretty clean way to do this where all the parsing code is extracted into its own internal function, and then we have two entrypoints, and the `uValue` complement is just performed on the return value for the function that handles negatives. Does that make sense? LMK and I can try the split myself.